### DISCUSSION TITLE
Picamera2 in python flask web app?

### DISCUSSION THREAD

**Discussion by krockamichael** [raspberrypi/picamera2#64] [Q&A] [answered]
Hi, I have just started using the picamera2 library and I would like to use it in my python flask web app. I am not sure how this would be possible, if at all. All I want to do is capture an image using the Picamera2() object and display it in an html template. This does not seem to work for me as the flask app "freezes" while the Picamera2 is initializing. This is my example code:

```{python}
@main.route('/')
def index():
  picam2 = Picamera2()
  capture_config = picam2.still_configuration()
  picam2.configure(capture_config)
  print('it is on!')
  image = picam2.capture_image()
  print('captured!')
  picam2.stop()
  return render_template('index.html', image=image)
```

After loading the home page of the app I see this in the terminal:

> [0:50:16.482520898] [3018]  INFO IPAManager ipa_manager.cpp:138 libcamera is not installed. Adding '/home/michael/libcamera/build/src/ipa' to the IPA search path
> [0:50:16.483172317] [3018]  INFO Camera camera_manager.cpp:293 libcamera v0.0.0+3495-bd38def2
> [0:50:16.485994508] [3020]  WARN CameraSensorProperties camera_sensor_properties.cpp:163 No static properties available for 'imx477'
> [0:50:16.486256079] [3020]  WARN CameraSensorProperties camera_sensor_properties.cpp:165 Please consider updating the camera sensor properties database
> [0:50:16.486441262] [3020] ERROR CameraSensor camera_sensor.cpp:591 'imx477 10-001a': Camera sensor does not support test pattern modes.
> [0:50:16.506693354] [3020]  INFO RPI raspberrypi.cpp:1340 Registered camera /base/soc/i2c0mux/i2c@1/imx477@1a to Unicam device /dev/media3 and ISP device /dev/media0
> 2022-04-11T19:35:34.936Z | INFO     | Initialization successful.
> 2022-04-11T19:35:34.936Z | INFO     | Camera now open.
> [0:50:16.509715061] [3018]  INFO Camera camera.cpp:1028 configuring streams: (0) 4048x3040-XBGR8888
> [0:50:16.510212241] [3020]  INFO RPI raspberrypi.cpp:757 Sensor: /base/soc/i2c0mux/i2c@1/imx477@1a - Selected sensor format: 4056x3040-SBGGR12_1X12 - Selected unicam format: 4056x3040-pBCC
> 2022-04-11T19:35:34.940Z | INFO     | Configuration successful!
> it is on!

But I never see the `print('captured!')` statements being reached, so I am forced to shut down the app. I have read some articles where the author used threading, but I am not very proficient in that area myself. To make things short (haha), is it possible to run picamera2 along with a python flask web app?

**Comment by krockamichael** [comment, marked as answer]
Update: I have found a way to simulate the [mjpeg_server.py](https://github.com/raspberrypi/picamera2#mjpeg_serverpy) in a flask app environment and it works! I just have to call picam2.stop_recording() every time I go to a different URL and then restart the picam2.

**Comment by mastrolinux** [comment]
I just did something that could be useful for your use case. I decoupled the image capturing from the flask presentation.
Here is the client code https://github.com/mastrolinux/raspberry-pi-security-camera-client and here is the server https://github.com/mastrolinux/raspberry-pi-security-camera-server Please let me know if it is useful for you as well. I am going to write an article about it.
